baselines/moon/moon/client.py

```diff
@@ -0,0 +1,159 @@
+"""Define your client class and a function to construct such clients.
+
+Please overwrite `flwr.client.NumPyClient` or `flwr.client.Client` and create a function
+to instantiate your client.
+"""
+
+import copy
+import os
+from collections import OrderedDict
+from typing import Callable, Dict, List, Tuple
+
+import flwr as fl
+import torch
+from flwr.common.typing import NDArrays, Scalar
+from omegaconf import DictConfig
+from torch.utils.data import DataLoader
+
+from moon.models import init_net, train_fedprox, train_moon
+
+# pylint: disable=E1101
+
+
+# pylint: disable=too-many-instance-attributes
+class FlowerClient(fl.client.NumPyClient):
+    """Standard Flower client for CNN training."""
+
+    def __init__(
+        self,
+        # net: torch.nn.Module,
+        net_id: int,
+        dataset: str,
+        model: str,
+        output_dim: int,
+        trainloader: DataLoader,
+        valloader: DataLoader,
+        device: torch.device,
+        num_epochs: int,
+        learning_rate: float,
+        mu: float,
+        temperature: float,
+        model_dir: str,
+        alg: str,
+    ):  # pylint: disable=too-many-arguments
+        self.net = init_net(dataset, model, output_dim)
+        self.net_id = net_id
+        self.dataset = dataset
+        self.model = model
+        self.output_dim = output_dim
+        self.trainloader = trainloader
+        self.valloader = valloader
+        self.device = device
+        self.num_epochs = num_epochs
+        self.learning_rate = learning_rate
+        self.mu = mu  # pylint: disable=invalid-name
+        self.temperature = temperature
+        self.model_dir = model_dir
+        self.alg = alg
+        # self.prev_net = init_net(self.dataset, self.model, self.output_dim)
+        self.prev_net = None
+
+    def get_parameters(self, config: Dict[str, Scalar]) -> NDArrays:
+        """Return the parameters of the current net."""
+        return [val.cpu().numpy() for _, val in self.net.state_dict().items()]
+
+    def set_parameters(self, parameters: NDArrays) -> None:
+        """Change the parameters of the model using the given ones."""
+        params_dict = zip(self.net.state_dict().keys(), parameters)
+        state_dict = OrderedDict({k: torch.from_numpy(v) for k, v in params_dict})
+        self.net.load_state_dict(state_dict, strict=True)
+
+    def fit(
+        self, parameters: NDArrays, config: Dict[str, Scalar]
+    ) -> Tuple[NDArrays, int, Dict]:
+        """Implement distributed fit function for a given client."""
+        self.set_parameters(parameters)
+        if self.prev_net is None:
+            self.prev_net = init_net(self.dataset, self.model, self.output_dim)
+            self.prev_net = copy.deepcopy(self.net)
+        else:
+            # load previous model from model_dir
+            self.prev_net.load_state_dict(
+                torch.load(
+                    os.path.join(self.model_dir, str(self.net_id), "prev_net.pt")
+                )
```

Comment:
Flower clients in simulation are ephemeral. This means that all their internal variables etc are discarded once the client completes the tasks it was spawned for (in this case to do `fit()`). This means that all the time, the above `else` is never happening (since `self.prev_net` is always None). I'd advice to change the if/else and instead load the model if a file is found with the matching name. Does it make sense?

Built-in support for stateful clients in Flower is almost ready. So later (after the _Summer of Reproducibility_ we'll update this logic)